core/src/main/java/org/sonarsource/sonarlint/core/SonarLintClientImpl.java

```diff
@@ -19,126 +19,242 @@
+    try {
+      LoggingConfigurator.setVerbose(verbose);
+    } finally {
+      rwl.writeLock().unlock();
+    }
+  }
+
+  public GlobalContainer getGlobalContainer() {
+    return globalContainer;
+  }
+
+  public void start() {
+    rwl.writeLock().lock();
+    this.globalContainer = GlobalContainer.create(globalConfig);
     try {
       globalContainer.startComponents();
+      if (globalContainer instanceof StorageGlobalContainer) {
+        changeState(((StorageGlobalContainer) globalContainer).getSyncStatus() != null ? State.SYNCED : State.NOT_SYNCED);
+      }
     } catch (RuntimeException e) {
-      throw handleException(e);
+      changeState(State.UNKNOW);
+      throw SonarLintWrappedException.build(e);
+    } finally {
+      rwl.writeLock().unlock();
     }
-    this.started = true;
   }
 
   @Override
   public RuleDetails getRuleDetails(String ruleKey) {
-    checkStarted();
-    return globalContainer.getRuleDetails(ruleKey);
-  }
-
-  @Override
-  public Collection<String> getActiveRuleKeys() {
-    checkStarted();
-    return globalContainer.getActiveRuleKeys();
+    rwl.readLock().lock();
+    try {
+      return globalContainer.getRuleDetails(ruleKey);
+    } finally {
+      rwl.readLock().unlock();
+    }
   }
 
   @Override
   public AnalysisResults analyze(AnalysisConfiguration configuration, IssueListener issueListener) {
     checkNotNull(configuration);
-    Preconditions.checkNotNull(issueListener);
-    checkStarted();
+    checkNotNull(issueListener);
+    rwl.readLock().lock();
     try {
       return globalContainer.analyze(configuration, issueListener);
     } catch (RuntimeException e) {
-      throw handleException(e);
+      throw SonarLintWrappedException.build(e);
+    } finally {
+      rwl.readLock().unlock();
     }
   }
 
-  private void checkStarted() {
-    if (!started) {
-      throw new IllegalStateException("SonarLint Engine is not started");
+  @Override
+  public GlobalSyncStatus getSyncStatus() {
+    checkConnectedMode();
+    rwl.readLock().lock();
+    try {
+      return ((StorageGlobalContainer) globalContainer).getSyncStatus();
+    } finally {
+      rwl.readLock().unlock();
     }
   }
 
-  private static RuntimeException handleException(RuntimeException t) {
-    if (LOG.isDebugEnabled()) {
-      // In DEBUG mode always return full exception
-      return convertToSonarLintException(t);
+  @Override
+  public void sync(ServerConfiguration serverConfig) {
+    checkNotNull(serverConfig);
+    checkConnectedMode();
+    rwl.writeLock().lock();
+    stop();
+    changeState(State.SYNCING);
+    ConnectedContainer connectedContainer = new ConnectedContainer(globalConfig, serverConfig);
+    try {
+      connectedContainer.startComponents();
+      connectedContainer.sync();
+    } catch (RuntimeException e) {
+      throw SonarLintWrappedException.build(e);
+    } finally {
+      try {
+        connectedContainer.stopComponents(false);
+      } catch (Exception e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)